File: java/insecure_deserialization/SessionHandling.java
import com.thoughtworks.xstream.XStream;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.Constructor;
import jakarta.servlet.http.*;
import java.io.*;
import java.util.Base64;

public class SessionHandling extends HttpServlet {

    // VULN 1: XStream.fromXML() on user-supplied XML input (RCE via deserialization)
    public Object restoreSessionFromXml(HttpServletRequest req) {
        String xmlPayload = req.getParameter("session_xml");
        XStream xstream = new XStream();
        return xstream.fromXML(xmlPayload);
    }

    // VULN 2: SnakeYAML constructor on user-supplied YAML (allows arbitrary class instantiation)
    public Object loadUserPreferencesFromYaml(HttpServletRequest req) {
        String yamlInput = req.getParameter("preferences_yaml");
        Yaml yaml = new Yaml(new Constructor());
        return yaml.load(yamlInput);
    }

    // VULN 3: Deserializing base64-encoded Java object from request parameter
    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException, ClassNotFoundException {
        String encodedSession = req.getParameter("session_obj");
        byte[] sessionBytes = Base64.getDecoder().decode(encodedSession);
        ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(sessionBytes));
        Object sessionData = ois.readObject();
        resp.getWriter().println("Session loaded for: " + sessionData.toString());
    }
}

File: kotlin/insecure_deserialization/SessionHandling.kt
import com.thoughtworks.xstream.XStream
import jakarta.servlet.http.HttpServletRequest
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.yaml.snakeyaml.Yaml
import java.io.ByteArrayInputStream
import java.io.ObjectInputStream
import java.util.Base64

@RestController
class SessionHandling {

    // VULN 1: SnakeYAML deserialization of user-controlled input - session preferences import
    @PostMapping("/api/session/import-preferences")
    fun importSessionPreferences(@RequestBody yamlInput: String): Map<String, Any> {
        val yaml = Yaml()
        val preferences = yaml.load<Any>(yamlInput)
        return mapOf("status" to "imported", "preferences" to (preferences ?: emptyMap<String, Any>()))
    }

    // VULN 2: XStream deserialization of user-supplied XML - order template import
    @PostMapping("/api/orders/import-template")
    fun importOrderTemplate(@RequestBody xmlInput: String): Map<String, Any> {
        val xstream = XStream()
        val orderTemplate = xstream.fromXML(xmlInput)
        return mapOf("status" to "imported", "template" to orderTemplate)
    }

    // VULN 3: ObjectInputStream from base64-decoded session param - session restore
    @PostMapping("/api/session/restore")
    fun restoreSession(
        @RequestParam("sessionData") sessionData: String,
        request: HttpServletRequest
    ): Map<String, Any> {
        val decoded = Base64.getDecoder().decode(sessionData)
        val ois = ObjectInputStream(ByteArrayInputStream(decoded))
        val sessionObj = ois.readObject()
        ois.close()
        val session = request.getSession(true)
        session.setAttribute("restoredData", sessionObj)
        return mapOf("status" to "restored")
    }
}
